core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -0,0 +1,307 @@
+
+import com.yammer.metrics.core.Meter
+import kafka.metrics.KafkaMetricsGroup
+import kafka.network.Processor.ListenerMetricTag
+import kafka.server.KafkaConfig
+import kafka.utils.TestUtils
+import org.apache.kafka.common.network._
+import org.apache.kafka.common.utils.MockTime
+import org.junit.Assert._
+import org.junit._
+import org.scalatest.Assertions.intercept
+
+import scala.jdk.CollectionConverters._
+import scala.collection.{Map, mutable}
+import scala.concurrent.TimeoutException
+
+class ConnectionQuotasTest {
+  private val time = new MockTime
+  private val listeners = Map(
+    "EXTERNAL" -> ListenerDesc(new ListenerName("EXTERNAL"), InetAddress.getByName("192.168.1.1")),
+    "ADMIN" -> ListenerDesc(new ListenerName("ADMIN"), InetAddress.getByName("192.168.1.2")),
+    "REPLICATION" -> ListenerDesc(new ListenerName("REPLICATION"), InetAddress.getByName("192.168.1.3")))
+  private val blockedPercentMeters = mutable.Map[String, Meter]()
+  private val knownHost = InetAddress.getByName("192.168.10.0")
+  private val unknownHost = InetAddress.getByName("192.168.2.0")
+
+  case class ListenerDesc(listenerName: ListenerName, defaultClientIp: InetAddress) {
+    override def toString: String = {
+      s"(listener=${listenerName.value}, client=${defaultClientIp.getHostAddress})"
+    }
+  }
+
+  def brokerPropsWithDefaultConnectionLimits: Properties = {
+    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, port = 0)
+    props.put(KafkaConfig.ListenersProp, "EXTERNAL://localhost:0,REPLICATION://localhost:1,ADMIN://localhost:2")
+    // ConnectionQuotas does not limit inter-broker listener even when broker-wide connection limit is reached
+    props.put(KafkaConfig.InterBrokerListenerNameProp, "REPLICATION")
+    props.put(KafkaConfig.ListenerSecurityProtocolMapProp, "EXTERNAL:PLAINTEXT,REPLICATION:PLAINTEXT,ADMIN:PLAINTEXT")
+    props
+  }
+
+  @Before
+  def setUp(): Unit = {
+    // Clean-up any metrics left around by previous tests
+    TestUtils.clearYammerMetrics()
+
+    listeners.keys.foreach { name =>
+        blockedPercentMeters.put(name, KafkaMetricsGroup.newMeter(
+          s"${name}BlockedPercent", "blocked time", TimeUnit.NANOSECONDS, Map(ListenerMetricTag -> name)))
+    }
+  }
+
+  @After
+  def tearDown(): Unit = {
+    TestUtils.clearYammerMetrics()
+    blockedPercentMeters.clear()
+  }
+
+  @Test
+  def testFailWhenNoListeners(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    val executor = Executors.newSingleThreadExecutor
+    try {
+      // inc() on a separate thread in case it blocks
+      val externalListener = listeners("EXTERNAL")
+      executor.submit((() =>
+        intercept[RuntimeException](connectionQuotas.inc(externalListener.listenerName, externalListener.defaultClientIp, blockedPercentMeters("EXTERNAL")))): Runnable
+      ).get(5, TimeUnit.SECONDS)
```

Comment:
* I think that `intercept` does not assert that the exception is thrown but return the intercepted exception. It would be better to use `assertThrown` here.
* Do we really need to use an executor here? It seems that `inc` should throw immediately if the listener does not exist.

Reply:
`intercept` also verifies the exception. For example, if the exception thrown is different, the test will fail with an error like this: `Expected exception java.lang.IllegalArgumentException to be thrown, but kafka.network.TooManyConnectionsException was thrown`. If if no exception is thrown, it will fail like this: `Expected exception java.lang.IllegalArgumentException to be thrown, but no exception was thrown`. From what I see, assertThrown achieves the same behavior as intercept.

Reply:
About using the executor -- you are right that `inc` should return immediately in this case. I still used the executor in case the behavior changes and `inc` blocks. I think it's better to protect the test from this scenario